geoutils/filters.py

```diff
@@ -0,0 +1,296 @@
+#
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Filters to remove outliers and reduce noise in rasters."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from typing import Any, Literal
+
+import numpy as np
+import scipy
+from scipy.ndimage import generic_filter as scipy_generic_filter
+
+from geoutils._typing import NDArrayNum
+
+try:
+    from numba import jit, prange
+
+    _has_numba = True
+except ImportError:
+    _has_numba = False
+
+    def jit(*args: Any, **kwargs: Any) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
+        """
+        Fake jit decorator if numba is not installed
+        """
+
+        def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+            return func
+
+        return decorator
+
+
+def _filter(array: NDArrayNum, method: str | Callable[..., NDArrayNum], **kwargs: dict[str, Any]) -> NDArrayNum:
+    """
+    Apply a filter to the array. See description of raster.filter.
+    """
+
+    if isinstance(method, str):
+        filter_map: dict[str, Callable[..., NDArrayNum]] = {
+            "gaussian": gaussian_filter,
+            "median": median_filter,
+            "mean": mean_filter,
+            "max": max_filter,
+            "min": min_filter,
+            "distance": distance_filter,
+        }
+
+        if method not in filter_map:
+            raise ValueError(f"Unsupported filter method '{method}'. " f"Available methods: {list(filter_map.keys())}")
+
+        filter_func = filter_map[method]
+
+    elif callable(method):
+        filter_func = method
+    else:
+        raise TypeError("`method` must be a string or a callable.")
+
+    # Apply filter
+    filtered_data = filter_func(array, **kwargs)
+    return filtered_data
+
+
+def gaussian_filter(array: NDArrayNum, sigma: float, **kwargs: Any) -> NDArrayNum:
+    """
+    Apply a Gaussian filter to a raster that may contain NaNs.
+    N.B: kernel_size is set automatically based on sigma.
+
+    :param array: The input array to be filtered.
+    :param sigma: The sigma of the Gaussian kernel
+
+    :returns: The filtered array (same shape as input)
+    """
+
+    if array.ndim == 1:
+        raise ValueError("Gaussian filter can't be applied to 1D arrays.")
+
+    # Boolean mask: True where NaN
+    mask = np.isnan(array)
+    mask = np.asarray(mask, dtype=bool)
+    # Replace NaNs with 0 for convolution
+    arr_filled = np.where(mask, 0, array)
+
+    # Apply gaussian filter to values and to mask
+    filtered = scipy.ndimage.gaussian_filter(arr_filled, sigma, **kwargs)
```

Comment:
Unless I miss something, there is an issue here! 
It is not enough to set the nans to 0. The filter must also be applied to the mask array, to propagate the number of valid values. See this [good example](https://stackoverflow.com/questions/18697532/gaussian-filtering-a-image-with-nan-in-python). 
For example, for array [1, 2, nan, 4], the mean is:
- with your approach: (1 + 2 + 4) / 4 = 7/4
- with the correct implementation: (1 + 2 + 4) / 3 = 7/3